pyth-sdk/src/price.rs

```diff
@@ -89,6 +90,94 @@ impl Price {
         self.div(quote)?.scale_to_exponent(result_expo)
     }
 
+    /// Get the valuation of a collateral position according to:
+    /// 1. the net amount deposited (across the protocol)
+    /// 2. the max amount depositable (across the protocol)
+    /// 3. the initial and final valuation discount rates
+    /// 
+    /// We use a linear interpolation between the the initial and final discount rates,
+    /// scaled by the proportion of max depositable amount that has been deposited.
+    /// This essentially assumes a linear liquidity cumulative density function,
+    /// which has been shown to be a reasonable assumption for many crypto tokens in literature.
+    /// 
+    /// Args
+    /// deposits: u64, quantity of token deposited in the protocol
+    /// max_deposits: u64, max quantity of token that can be deposited in the protocol
+    /// discount_initial: u64, initial discount rate at 0 deposits (units given by discount_precision)
+    /// discount_final: u64, final discount rate at max_deposits deposits (units given by discount_precision)
+    /// discount_precision: u64, the precision used for discounts
+    /// 
+    /// Logic
+    /// collateral_valuation_price = (deposits * price * (discount_precision-discount_final) + (max_deposits - deposits) * price * (discount_precision - discount_initial)) / (max_deposits * discount_precision)
+    pub fn get_collateral_valuation_price(&self, deposits: u64, max_deposits: u64, discount_initial: u64, discount_final: u64, discount_precision: u64) -> Result<Price, LiquidityOracleError> {
+        if max_deposits < deposits {
+            return Err(LiquidityOracleError::ExceedsMaxDeposits.into());
+        }
+
+        if discount_initial > discount_final {
+            return Err(LiquidityOracleError::InitialDiscountExceedsFinalDiscount.into());
+        }
+
+        if discount_final > discount_precision {
+            return Err(LiquidityOracleError::FinalDiscountExceedsPrecision.into());
+        }
+
+        let remaining_depositable = (max_deposits-deposits) as i64;
```

Comment:
you should use `i64::try_from` instead of `as` if there's a chance the cast can fail. In this case, the max u64 can't be casted to an i64 `as` will convert it into a negative number. (you may also be able to do `number.try_into()` -- i forget if that's implemented for numbers)

Reply:
agreed, this is a good call. fwiw i think this should be done across the sdk, a lot of instances of ___ as i64 in the sdk code rn